### DISCUSSION TITLE
Are contributors to blame for the oversight by maintainers?

### DISCUSSION THREAD

**Discussion by sktpy** [checkstyle/checkstyle#14621] [General]
Refer to this conversation ([Link](https://github.com/checkstyle/checkstyle/pull/14553#discussion_r1514865757))

The maintainer in question seems to have missed where we have tested the new feature,
which is _totally fine_, mistakes happen, sometimes things inadvertently are able to slide through.

So I sent the links two the two files relevant to the test.
I don't have any problem till this part.

Note that these two inputs were specifically added to test the feature in question,
From top to bottom, each and every character in there, is to test the "specific" in question.

But then, maintainer in question proceeds to provide advice to "to include line numbers".

That does not makes any sense at all, as the whole two files, each and every character, only contains what is relevant to testing the new feature.

Including line numbers can be helpful where:
- I have extended some existing test, where some may not contain relevant parts.
- I am asked for some specific case, like test for where it is in anonymous class etc.
For things above that totally makes sense, to include line number/or permalink to the code selection for the ease of review, and I have done so in the past.

But then I thought, maybe the maintainer in question missed that those whole files contain only the relevant test inputs,
which I then explained in the following [comment](https://github.com/checkstyle/checkstyle/pull/14553#discussion_r1515388818).
And even [asked](https://github.com/checkstyle/checkstyle/pull/14553#discussion_r1515397719) to reconsider if this was suitable for this specific case.

_Everything is fine to me till here_, it is what the comments that maintainer in question sends afterwards that is really **_distasteful_ and _disrespectful_** at best, and "**gaslighting**" at worst.
It's one thing to miss something, but it is another thing to not accept the oversight and then double down on the delusional statements.

> I wouldn't have mentioned it if I didn't think it would be helpful. This item is resolved.

Do we now have to state obvious facts in the PR? 
Is anyone here confused the only input file that were created as the part of PR contains the relevant inputs and not the secret recipe for Coca-Cola?

While we the contributors are reminded about that how busy maintainers are and we shouldn't waste their time,
What is it that makes the maintainer in question believe that contributors have got time for unemployed behaviour either?

If any one believes that I am blowing this out of proportions, please consider the effort that is put out and when someone does not have enough decency to even look at the changes before shooting out illogical advice, it is really hurtful.

What I am expecting here is either of the two as following:
- Either the maintainer in question can accept the oversight.
- Or, because we cannot force anyone to do anything, what I would like if to gather if the other maintainers feels the same way the maintainer in question does.

I believe this is an isolated incident and does not reflects the view of this project and other people involved in this, or even the maintainer in question, I do believe the maintainer in question had the best interests in the mind.
I always go above and beyond to provide explanations, relevant links in the PR for the ease of the maintainers,
along with that I am *always* open to any suggestion, nrmancuso can probably confirm this.

I am contributing foremost for the joy of contributing, and love for computer science, yes being selected for gsoc is nice but that's not the end of the world, I am looking for projects that I can be part of for long term.
But if this is the collective thinking of all the maintainers that they somehow are superior, and require to be spoon fed (which I highly believe is not the case for all), I cannot continue here, so please do tell me if that is the message we are going for.

But in that highly unlikely scenario, I would still finish any PR and Issue that concerns me, and in that highly unlikely case please have my sincere gratitude for providing a welcoming place for new contributors, and taking the time to review my PRs.

Thank you.

**Comment by sktpy** [comment]
cc @nrmancuso

**Comment by nrmancuso** [comment]
@sktpy thank you for sharing your thoughts about this. Asynchronous communication over the internet, across time zones, languages, and cultures is *hard*. This is something that I have learned (and are continually reminded of) over the last few years as I have become active in Open Source and moved to a fully remote position in my professional career. I try to always imagine that the other person has the absolute best intentions, which you've mentioned above.

>While we the contributors are reminded about that how busy maintainers are and we shouldn't waste their time

No one should be "reminding" anyone of this, because it is a falsehood. We are all busy humans, trying to do our best.

>blame

I don't think we need to place blame on anyone here, and I am not sure how it would help. Different people have different communication styles, and as technical people, we tend to be more direct.

Would sharing a link to a specific line have been more helpful? Of course. Was it necessary? IMO, no. 

>But if this is the collective thinking of all the maintainers that they somehow are superior

This is absolutely not my thinking, and if this is the message we are conveying, then we need to do better.

**Comment by sktpy** [comment]
@nrmancuso  Thank you for going through my rant. :heart: 
<hr><br>

> Asynchronous communication over the internet, across time zones, languages, and cultures is hard.

That's a great point.
<hr><br>

> I try to always imagine that the other person has the absolute best intentions

What's in good sport can still be hurtful nonetheless, I guess that will just start to subdue as one gets more experienced.
<hr><br>

> No one should be "reminding" anyone of this

Sad reality is we do this in one way or other, this hasn't been the case for me, but I have seen this done to other contributors, iykyk.
Though I couldn't help but notice the fault of contributors in those cases, either if it is pinging at 60 frames per second, or failing to do their homework.
But we should be more professional on how we deal with such cases in a way that is non condescending.

**Edit:** That's a really admirable trait of you to believe that it is a falsehood :star: :heart: 🫡, not everyone is in the same boat.
I can't beleive that I am saying this but for the live example of such reminder refer to the second paragraph of [this](https://github.com/checkstyle/checkstyle/discussions/14621#discussioncomment-8730118) reply to this comment, somehow heavy work day, personal errands, problems are the things that apparently only a maintainer can go through. Everyone else is immune to such things I guess.

It isn't realistic to expect that everyone can think the same way, but I am glad that I have atleast one shoulder to cry on. :slightly_smiling_face: 
<hr><br>

> This is absolutely not my thinking

I hope everyone shares the same feeling :)
<hr><br>

> and if this is the message we are conveying, then we need to do better.

We do need to, presenting report shouldn't feel like we are presenting it to some head of state.
With that convo it seems like contributor must first do all the homework, and then must spoon feed each and every bit.
<hr><br>

I can only speak for myself, but I put extra care in every issue and PR to ease the review process in any way possible.

Because at the end of the day, maintainers do need to handle a large variety of things, no one is denying that,
and every reasonable help that can be provided to ease this should be done.

Just because one can ask for something, doesn't mean they should, we should be thoughtful about what is asked for.
<hr><br>

At the end I would like to conclude this with the quote from romani,
> "We like student that don't need much assistance"

That's a really great motto, wouldn't it even be better if both parties can extend the same diligence?
Let's be the role model for what we want to see in others, because after all we are all monkeys, monkey see, monkey do.

I take great care, in return I don't expect any more than what would eliminate the need for spoon feeding, I hope it is not too much to ask for.

:heart:

**Reply by romani** [reply]
> That does not makes any sense at all,

open source communication is complicated as people usually contribute not in there best time or energy level.
and it is also complicated due to written style, you can find a lot of materials why written text sometime considered to be rude. 
One more complication is amount of tickets and conversations we have to do to complete some task, numerous of contributors. close to 100 not completed conversation in PRs, ........ Context switching if very complicated process for maintainers, lack of clarity in conversation/wording make it even worse.

Maintainers usually have random 10 minutes, after heavy work day and tons of personal errands and problems. Instead of spending of this 10 minutes on youtube, they do last effort to make some responses in PRs. So sometime maintainers do not have time and enough energy to go very deep in PR and read each line. Due to proportion of 10-30 contributors to 1 maintainer, we fall in model where maintainers do less deep focus activities but to respond/guide more contributors.
So we think, that some feedback (may be it will be considered as lazy) is  better then no feedback at all for long time (when maintainer try to find slot of time to give you very comprehensive and thoughtful feedback).

Keep in mind that maintainers keep doing project for many years, few of us, do it for more than 10 years, I hope you can understand that level of energy of new contributor and 10 years are dramatically different.

You never know in what situation maintainer give response, sometime we have only few minutes, or small screen(we did a lot to force dev process to friendly to review from phone screen)  or one finger to send you a message. 

If you see that maintainers are "lazy" to do something them self of should do more efforts themself, .... keep this picture in mind - https://twitter.com/checkstyle_java/status/1439339680047063040?t=EJflQoOXgPcXAiJRae_W4A&s=19 , it is real situation.
I was blamed numerous times that I am lazy and should do more. But not of people who blame run a project for long time :). Some PRs waiting for me for years, but I have no enough energy to re-read all and make very complicated design decision, so some PRs are stuck in unfortunate state. 

Please bare with us, sorry, lets keep moving forward this complicated opensource initiative together. While we are in project, we already do our best, if we ask for some details, please share it. If you do not want to share, it will just takes longer to accept PR. 
If non of side have time/energy to do a bit more efforts, it is ok, it will wait for time when some side have more time.

I rarely share same request to point me to exact line of code to help me accept PR quicker. Please help us.

**Reply by sktpy** [reply]
> energy of new contributor and 10 years are dramatically different.

That's a very reasonable point.
<hr><br>

> keep this picture in mind - https://twitter.com/checkstyle_java/status/1439339680047063040?t=EJflQoOXgPcXAiJRae_W4A&s=19

That is truly noble, but in my opinion a bit too noble, child is the priority here, these moments are too precious to be mingled with any other stuff. :heart: * 3000
<hr><br>

> if we ask for some details, please share it

Please consider the special circumstances of this case.

I don't have any issue with sharing details, but in that _particular_ scenario, pointing to the exact line would be fruitless because the config is needed to be kept in context to analyse if we are truly testing the feature correctly,
as the maintainer in question mentions: `"I can already see the file but not the specific cases"`

But I fail to see how can one not see the "specific cases" but see the file at the same time, when the whole file _is_ the "specific case".
That just didn't made any sense to me, which I again have pointed out further in the conversation but in vain.

The thing is that it isn't too much effort to just sent the link to the line, I fail to see how exactly it would have provided any help whatsoever, in this specific case only.

As I have mentioned, I understand that the workloads of  maintainers is really huge.

But if I am to fullfill such request, for such "specific" scenario, I would have to do it with my logical thinking turned _off_.
If that can help maintainers, I am all for it, it isn't really much of effort for me in the first place, so why not.
<hr><br>

> Maintainers usually have random 10 minutes, after heavy work day and tons of personal errands and problems.

Thanks for the reminder regarding the "busy nature of maintainers".  As I have mentioned, no one is denying that. :)
<hr><br>

But you would have to understand why a new contributor like me might get frustrated over this, being unknown to this workflow,
a thing which one can only understand if they themselves go through the phase,
so I just wanted to clear out what was on my mind, as this internal frustration eventually would have killed the joy for me, and I didn't really wanted it to happen.

Apologies for any of my statements that weren't too civil.

I hope you understand.
Thank you.

**Reply by romani** [reply]
>But I fail to see how can one not see the "specific cases"

Please do share link, maintainer might be lost, or lost track and asking you to guide him . Please help us.

>would have to understand why a new contributor like me might get frustrated over this,

I know, I had this numerous times.we all people, some time we need to do job being under pain and tiredness, bad mood. If it happens, take a break and maybe next day you will read post in different way or agree with opponent. Maintainers are not gods they also need sometime guidance.

**Comment by sktpy** [comment]
I am glad that we were able to have a discussion regarding this, and clear the things out.

I believe it may be helpful in providing insights to the budding contributors, as I believe these thoughts may have crossed various minds,
but not everyone is stupid enough like me to ruin their chances here. :joy: 

But I hope we can look past all that and keep the spirit alive. :heart:

**Comment by sktpy** [comment]
@romani I have been thinking about your [comment](https://github.com/checkstyle/checkstyle/discussions/14621#discussioncomment-8730118), here are my thoughts.

**Disclaimer:** This can wait, but if you got time, feel free to go through it.

While having discussion over this may seem a waste of time, but I believe that it is important to step back from technical things for a few moments and understand the social dynamics at play here, so that we all can partake in the initiative in a better way.
<hr><br>

> Maintainers usually have random 10 minutes, after heavy work day and tons of personal errands and problems. Instead of spending of this 10 minutes on youtube, they do last effort to make some responses in PRs.

That is very noble of them. But I believe in my opinion, someone who is not a maintainer can be going through the same as well.
Take a look at this https://github.com/checkstyle/checkstyle/issues/14618#issuecomment-1987370899, a response which I have given at 3 AM (my time), due to the busy day, but still decided to do that, eliminating an extra hour which I could have slept and go to the college in the morning with a little more energy.

But would I say that I sacrificed something here?
The answer for me would be, absolutely not.
Because for me, the joy and contentment I get from contributing simply just exceeds any activity that I could have partake in,
whether if it is catching on the episode of Shogun that was premiered last tuesday that I wanted to watch but haven't go the time yet to do that due to various things, or various lucrative fun things one can enjoy.

But if I may, I would like to give an unsolicited advice,
If some one rather feel more joy from watching 10 minutes of youtube instead of responding to PR,
In my opinion, they should spend that time on youtube.
These PR stuff are secondary and can wait for eternity, we should focus on what provide us joy because life is too short,
everything else can wait, PRs aren't going anywhere. :slightly_smiling_face:

> keep this picture in mind - https://twitter.com/checkstyle_java/status/1439339680047063040?t=EJflQoOXgPcXAiJRae_W4A&s=19

And in the cases like above, if only situation in which one can response to PR is via multitasking when they push their child on the swing.
They should absolutely be not handling any PR or whatever, these are the things that can wait.

But somehow, if one still feels obliged to respond to PR and whatnot even in those moments, first of all huge respect to them,
but no one should be reminded about these things, because these are things that anyone can be subjected to, and is not exclusive to the maintainers only (in my limited experience regarding the life, I can be wrong). Everyone of us are human beings.

<hr><br>

> energy of new contributor and 10 years are dramatically different.

This is a very reasonable point and I agree 100%, this can explain many things.

I don't have any problem whatsoever fulfilling any request that can help maintainers.

I can understand that they may be in rush, having a bad day, or just lost for some reason, and even in those cases if they still go on to handle theses issues and PR, that is some serious dedication, and words cannot express my gratitude.

But at the end of the day, I, we the contributors are also human beings and hence unfortunately subjected to human emotions.
If I had ability of become a rock in a river that is unperturbed by the gushing streams of water, and that can help me serve the cause in a better way, I would have, but I haven't yet learnt such ways of tranquility, though I wish that I can achieve it some day.

So while from now on I will keep the pain of maintainers in mind,
if possible my only humble request here is that to not label [such](https://github.com/checkstyle/checkstyle/pull/14553#discussion_r1514865757) asks as something "You should have".
These little things go a long way, as I am not perfect either, I am just a human being here, like everyone else.

I do understand the complicate nature of asynchronous communication, wording and what not, energy level and all the bells and whistles.
But whenever possible, let's be atleast polite about such requests and not label it as something that one should have done.

While I can remember this now, and it is highly likely that I would never be startled by such things,
there are many who won't have gone through this, and quite possibly too afraid to speak up anything about it but at same time might be having internal frustration regarding the same which is no good.
Let's do what we can to be more welcoming to the cause.

:heart:

**Reply by romani** [reply]
>They should absolutely be not handling any PR or whatever, these are the things that can wait.

:). You will understand this in 20 years of hard working on limited resources.
All that you see around is because collaboration happens even there is no real time for keyboard. 

This is my last comment. 
You decide yourself to help maintainers or not.

**Reply by sktpy** [reply]
Thanks for responding.

> You will understand this in 20 years of hard working on limited resources.

I hope I will. :)

**Comment by sktpy** [comment]
@nrmancuso ~~This can wait but if it is not too much of trouble, I would like to hear your views on https://github.com/checkstyle/checkstyle/discussions/14621#discussioncomment-8741560 whenever you get time to do so. :slightly_smiling_face:~~

**Edit:** Nvm, Ignore this comment, didn't felt right to delete it since I have pinged you.

I have learned some valuable lessons here regarding interaction in oss, will post them once I have given them deep thought, so that it may benefit anyone who wanders here and is going through the same.
